### PR TITLE
safe_url_string, canonicalize_url: apply stripping from the URL living standard

### PR DESCRIPTION
**Pull request by Gallaecio**

### CODE REVIEW

**Review by kmike — APPROVED** [review]
Looks good, thanks @Gallaecio!

Would it make sense to update safe_url_string docstring? It's no longer only tabs and newlines which are removed.

File: w3lib/url.py [hunk 1]
```diff
@@ -77,9 +77,18 @@ def _quote_byte(error: UnicodeError) -> Tuple[str, int]:
 _FRAGMENT_SAFEST_CHARS = _PATH_SAFEST_CHARS
 
 
-_ascii_tab_newline_re = re.compile(
-    r"[\t\n\r]"
-)  # see https://infra.spec.whatwg.org/#ascii-tab-or-newline
+_ASCII_TAB_OR_NEWLINE = "\t\n\r"
```
Comment (wRAR) [inline w3lib/url.py:80]: Doesn't this duplicate w3lib._infra?

**Comment by Gallaecio** [comment]
Is https://github.com/scrapy/w3lib/pull/207/commits/288caae88640e0243aed1a7f8e73047c639e324a OK? It describes more what the functions should do than what they actually do.

**Review by wRAR — APPROVED** [review]

File: w3lib/url.py [hunk 2]
```diff
@@ -531,11 +556,8 @@ def canonicalize_url(
 ) -> str:
     r"""Canonicalize the given url by applying the following procedures:
 
+    - make the URL safe (see :func:`safe_url_string`)
```
Comment (kmike) [inline w3lib/url.py:559]: That's actually a question I had :) It seems canonicalize_url doesn't call safe_url_string. How do we ensure it makes the URl safe in the same way? Or is it not a requirement? From the docstring I'd infer that canonicalize_url calls safe_url_string, and then does something on top of it.

File: w3lib/url.py [hunk 3]
```diff
@@ -88,9 +95,29 @@ def safe_url_string(  # pylint: disable=too-many-locals
...
+    """Return a URL equivalent to *url* that a wide range of web browsers and
+    web servers consider valid.
+
+    *url* is parsed according to the rules of the `URL living standard`_,
+    and during serialization additional characters are percent-encoded to make
+    the URL valid by additional URL standards.
+
+    .. _URL living standard: https://url.spec.whatwg.org/
+
+    The returned URL is valid by *all* of the following URL standards known to
```
Comment (kmike) [inline w3lib/url.py:107]: @Gallaecio I think you mentioned that the implementation is not 100% compatible, until we solve https://github.com/scrapy/w3lib/issues/204. Is there some detail to it?

Comment (Gallaecio) [reply w3lib/url.py:107]: No, there is not. This is why I said:

> It describes more what the functions should (in my opinion) do than what they actually do.

Any suggestion on how to handle this? Should we say that the function *aims* to do this, but there are know issues or divergences at the moment?

Comment (kmike) [reply w3lib/url.py:107]: A small language tweak ("should be valid" instead of "is valid") might be enough.

File: w3lib/url.py (hunk 2 above)
Comment (kmike) [reply w3lib/url.py:559]: If we revert, it won't be worse than before, but the docstring will not be correct - it mentions many things, but not the whitespace handlng. So, some options:

1. Revert the docstring change
2. Write "make the URL safe" without referring to safe_url_string
3. Document that whitespace and control characters are stripped

I'm fine with all 3 :)

**Comment by kmike** [comment]
Thanks @Gallaecio!

**Pull request merged**